### ISSUE TITLE
Question about Batch Normalization

### ISSUE THREAD

**Issue by Ang-Li-95** [pkomiske/EnergyFlow#16]
Hi,
I am thinking about using Batch Normalization to avoid overfitting when training an EFN/PFN model but I don't find anything relevant to that in your document.

I am wondering whether Batch Normalization is implemented in the package? If yes, could you please give me some instruction on how to use that? If not, would it be useful to have that in the EFN package in the future?

Thanks,
Ang

**Comment by pkomiske** [comment]
Hi Ang,

Batch Normalization is not currently implemented in the package, although in principle it could be since Keras supplies a `BatchNormalization` layer and the EnergyFlow architectures are all implemented using Keras. 

There seems to be [some debate](https://stackoverflow.com/questions/34716454/where-do-i-call-the-batchnormalization-function-in-keras) about exactly where Batch Normalization should go. I'm hesitant to implement things myself if I'm not planning on immediately using them because my bandwidth for thorough testing is limited. Do you have a concrete proposal for what you'd like to do? If so, feel free to try implementing it and making a pull request.

In practice, we've found that EFNs and PFNs don't suffer from very serious overfitting because the underlying network components can be quite small. As always, the best regularization is more data. Have you observed serious overfitting, and if so, what are the details of your training?

Best,
Patrick

**Comment by Ang-Li-95** [comment]
Hi Patrick,

Thank you for your kindly reply!

I asked about this just because it is a way to avoid overfitting that I want to see whether it is useful. I don't have robust reasoning for that.

And yes there is not serious overfitting in the EFN/PFN when I set some dropout or reg factor. Also, I used earlystopping to avoid overfitting. I am just curious about whether BN helps to get better results.

So it's totally fine if BN is not implemented in the package.

Best,
Ang

**Comment by pkomiske** [comment]
Hi Ang,

Glad to hear that it's not a terrible problem without BN. I've also been curious in the past if it's something that would help the training avoid overfitting or just generally be more robust. In case you want to try using BN, I'll describe below what you'd want to do:

1. Install EnergyFlow via the git repo. This means you probably want to first uninstall it, clone the repo (or your fork thereof), and then use pip to install from that directory:
```python
pip3 uninstall energyflow
git clone https://github.com/pkomiske/EnergyFlow.git # use your own fork here
cd EnergyFlow
pip3 install -e .
```
Now, any changes you make to the source code in that directory will be automatically taken into account the next time you import energyflow within python. Keep in mind python's strange import system means that you'll likely need to start a fresh instance or restart the kernel (if using jupyter) for a package to actually reload.

2. Import the `BatchNormalization` layer from Keras at the top of [efn.py](https://github.com/pkomiske/EnergyFlow/blob/master/energyflow/archs/efn.py). Then edit lines [113-120](https://github.com/pkomiske/EnergyFlow/blob/master/energyflow/archs/efn.py#L113-L120) and/or [156-171](https://github.com/pkomiske/EnergyFlow/blob/master/energyflow/archs/efn.py#L156-L171) to include the BN layer how you want. Try to ensure that any new layers/tensors you create are added to the appropriate list that stores all of the layers/tensors in the appropriate order.

3. If you want to be able to pass options to the BN layer from your code, take a look at how [the other options are stored](https://github.com/pkomiske/EnergyFlow/blob/master/energyflow/archs/efn.py#L253-L281) and add whatever you want for BN. Then pass those options to the functions that construct the main layers of [Phi](https://github.com/pkomiske/EnergyFlow/blob/master/energyflow/archs/efn.py#L253-L281) and/or [F](https://github.com/pkomiske/EnergyFlow/blob/master/energyflow/archs/efn.py#L373), which are where you added the BN layer earlier.

4. Test it out and make a pull request from your fork if it went well!

Best,
Patrick

**Comment by Ang-Li-95** [comment]
Hi Patrick,

That's very helpful. Currently, I am using EFN on a cluster so I am not sure whether I can uninstall things and modify that. 
But I may try your suggestion when I have time on that.

Thanks,
Ang

**Comment by pkomiske** [comment]
I'm closing this for now. Feel free to report any attempts in the future to implement Batch Normalization or other features if you want.
